Stand-alone client

AlexanderDzhoganov opened the pull request:
I've managed to parse the game database thanks to [this document](https://github.com/NimVek/pipboy/blob/master/PROTOCOL.md) by [NimVek](https://github.com/NimVek) (Huge thanks for writing that!). The JSON blob is pretty large and contains a lot of info. Here is an [example JSON dump](https://gist.github.com/AlexanderDzhoganov/edc0c0e7f38a21787cb8).

Also I've started work on a standalone client (`standalone.js`) as the mobile app makes my device get quite hot after a while and running the app for hours on end is painful.

~~Depends on https://github.com/rgbkrk/pipboylib/pull/10~~
~~Depends on https://github.com/rgbkrk/pipboylib/pull/14~~

AlexanderDzhoganov committed:
packet decoding + parse database

AlexanderDzhoganov committed:
add bufferutil dependency to package.json

AlexanderDzhoganov committed:
package.json fix

AlexanderDzhoganov committed:
standalone client

rgbkrk commented:
Oooh awesome! Yes, we definitely need a standalone client. The relay was put up as a way to start figuring the protocol out without having to mess with packet captures.

rgbkrk commented on cli.js:
```diff
@@ -1,60 +1,64 @@
-var pipboylib = require('pipboylib')
-var relay = require('pipboylib/lib/relay')
+var pipboylib = require('pipboylib');
+var relay = require('pipboylib/lib/relay');
```
Amusingly, a friend of mine lambasted me for going the no-semicolon route while another praised it.. I actually prefer them in, so I welcome this. There's even a [linter for this style](https://github.com/rgbkrk/obama) with a presidential seal of approval.

nelix replied:
I spent so much time removing semicolons :P I use them at work, it will take some time :P 
Before obama there was semistandard (part of the same group of projects as standard)

AlexanderDzhoganov replied:
I come from a C++ background and semicolons are _the way of my people_, though I am prepared to accept defeat and write w/o the semicolons if that's the majority decision :smiley:

rgbkrk replied:
Bring on the semi-colons!

I'm a bigger fan of them existing since that's how people tend to learn JavaScript and folks don't need to understand automatic semicolon insertion. Sorry about starting us off on a polar topic. :smile:

rgbkrk replied:
By the way, obama is all semistandard under the covers. I made it in frustration of dueling linter wars.

AlexanderDzhoganov committed:
client should reply to heartbeats + socket.io server

rgbkrk commented on pipdb.js:
```diff
@@ -0,0 +1,114 @@
+var dbTypes = {};
+
+var dbEntryType = ['Bool', 'Int8', 'UInt8', 'Int32', 'UInt32', 'Float', 'String', 'List', 'Dictionary'];
+
```
We actually have these setup in [pipboylib as a parser](https://github.com/rgbkrk/pipboylib/blob/master/lib/parser.js) thanks to @nelix. Can we bring more of this type of functionality there, supplementing where possible?

AlexanderDzhoganov replied:
I wasn't sure how you'd prefer to split up the functionality between the two projects. Feel free to split this up and move it around however you deem necessary.

rgbkrk replied:
Ok, always happy to `git cherry pick`.

Perhaps I should have outlined it in each. My intention was that bits that would be useful by any node application would go to `pipboylib` while `pipboyrelay` is one cli tool that uses them.

rgbkrk commented on pipdb.js:
```diff
@@ -0,0 +1,114 @@
+  } else {
+    return dbIndices[index];
+  }
+}
```
The databasing we definitely don't have anywhere. Big thanks to you and @nimvek!

rgbkrk commented on standalone.js:
```diff
@@ -0,0 +1,75 @@
+io.on('connection', function(socket) {
+  console.log('Socket.io client connected!');
+  socket.emit('db_update', db);
+});
```
socket io is a neat idea to use here

AlexanderDzhoganov committed:
standalone client should send heartbeat every 100ms

AlexanderDzhoganov committed:
standalone.js use pipboylib's discovery class

AlexanderDzhoganov committed:
refactor pipdb.js

AlexanderDzhoganov committed:
fixes

AlexanderDzhoganov commented:
I've started a repo for a browser-based client to the socket.io server - https://github.com/AlexanderDzhoganov/pipboyclient

Got real-time world map working - [screenshot here](http://i.imgur.com/oKNOt6h.png)

AlexanderDzhoganov committed:
don't crash if db decode fails

nelix commented:
This is very cool, I am stuck at work so I can't play with it yet.
Can't wait.

nelix commented on pipdecode.js:
```diff
@@ -0,0 +1,32 @@
+require('buffertools').extend();
```
have you seen https://github.com/rgbkrk/pipboylib/blob/master/lib/parser.js

You seem to have implemented way more protocol though!

The api I would eventually like to have would allow piping the relay into the parser into websocket/filters/whatever into serialiser into relay.

AlexanderDzhoganov replied:
I will try to merge my code with the pipboylib parser and make a PR to the other repo.

nelix replied:
If you like, I am happy to explore a few different directions.
We could try a few different angles on the parser and the cli.

rgbkrk replied:
:+1:

landongn commented on standalone.js:
```diff
@@ -0,0 +1,91 @@
+    process.exit(1);
+  });
+
+  pipdecode.onPacket = function(packet) {
```
might be worth having a `packetHandler` here that implements `EventEmitter`.  Might make things tidy as more and more channels get implemented. (nice work, btw!)

AlexanderDzhoganov replied:
Sounds good, I'll look into refactoring this a bit later today.

mattbaker commented:
> I've managed to parse the game database thanks to this document by NimVek (Huge thanks for writing that!). The JSON blob is pretty large and contains a lot of info. Here is an example JSON dump.

Hot damn. Well done.

rgbkrk commented:
Looking through both this PR and your really excellent client @AlexanderDzhoganov - it would make a lot of sense to make it be an Electron app. You wouldn't have to use socket.io (direct node-to-browser just fine in Electron) and user's won't be running servers on their machines, just a full native app they can download.

AlexanderDzhoganov committed:
clean-up, most code moved to PR in pipboylib

AlexanderDzhoganov commented:
I have done some clean-up and moved most of the code to [a separate PR in pipboylib](https://github.com/rgbkrk/pipboylib/pull/10). I have removed all of my changes to `cli.js` for the time being.

AlexanderDzhoganov committed:
remove changes to cli.js

landongn commented on standalone.js:
```diff
@@ -0,0 +1,87 @@
+  io.listen(SOCKETIO_PORT);
+  console.log('Socket.io server listening on port ' + SOCKETIO_PORT);
+
+  pipDecode.on('db_update', function(data) {
```
Awesome!  This looks much easier to manage.  Nice work! :+1:

AlexanderDzhoganov committed:
more clean-up

AlexanderDzhoganov committed:
initial work on client commands

rgbkrk commented:
Wow I finally looked at that JSON dump. Definitely huge. :smile:

rgbkrk commented:
Ok, yeah this is looking really nice. I went through and reimplemented this locally using pipdb and pipdecode and it made me realize where the relay fell short. If I connect, disconnect, connect, I end up putting the state of the last `expectedSize` out of whack, causing it all to crash.

rgbkrk commented:
Awesome. I think this is ready to at the very least be part of the current relay right now.

rgbkrk commented:
As in, I'm using this and it's pretty awesome.

![screenshot 2015-11-24 21 30 01](https://cloud.githubusercontent.com/assets/836375/11387838/8ec4e4cc-92f4-11e5-83d2-c112a1e921aa.png)

rgbkrk commented:
![screenshot 2015-11-24 21 51 43](https://cloud.githubusercontent.com/assets/836375/11387970/fb9ca502-92f5-11e5-9273-5448caea6d4f.png)

rgbkrk commented:
Definitely wondering if we should be doing some post processing of the db to turn all the arrays of `{value: 8000, text: "Caps Found", showIfZero: true}` into `"Caps Found": {value: 8000, showIfZero: true}`. This would be under the assumption that there are no collisions on the keys there.

Tradeoffs are that we're currently mirroring the actual format vs. something friendly to reach into programmatically in JavaScript (assuming order may change).

landongn commented:
@rgbkrk If you implement it as some kind of mapping, I could see that being nice.  Flipside though, if this ends up being used by other services expecting parity with the wire protocol, it could cause some divergence.  Maybe once there's specific needs?

rgbkrk commented:
Yeah, that's definitely the tradeoff. You'd think that if folks expect true parity with the wire protocol that we would just saw the raw bytes at that point. Way more compressed, certainly.

landongn commented:
Since the work for implementing `eventEmitters` is happening, that extra event would be a pretty easy way to grab it rather than the other events. 

:bike:

NimVek commented:
If you change the order or style of the Tree, you should remember that updates always have an index in the db. So you must always trace back how you changed ist. Better is a type of overlay to the db.
Second thing is that under sortedIDS, stimpackObjectID and radawayObjectID are indexes to the db are stored as values.

AlexanderDzhoganov committed:
resolve conflict in package.json

AlexanderDzhoganov committed:
Merge branch 'master' of https://github.com/AlexanderDzhoganov/pipboyrelay

AlexanderDzhoganov committed:
change client heartbeat timer to 1s instead of 100ms

AlexanderDzhoganov committed:
localmap support

AlexanderDzhoganov committed:
Merge branch 'master' of github.com:AlexanderDzhoganov/pipboyrelay

rgbkrk commented:
Now that rgbkrk/pipboylib#14 is merged, think this is ready?

AlexanderDzhoganov commented:
Yes, we should probably merge this now.
